File: 3598.longest-common-prefix-between-adjacent-strings-after-removals.java
#
# @lc app=leetcode id=3598 lang=java
#
# [3598] Longest Common Prefix Between Adjacent Strings After Removals
#

# @lc code=start
class Solution {
    private int commonPrefix(String a, String b) {
        int minLen = Math.min(a.length(), b.length());
        int i = 0;
        while (i < minLen && a.charAt(i) == b.charAt(i)) {
            i++;
        }
        return i;
    }

    public int[] longestCommonPrefix(String[] words) {
        int n = words.length;
        if (n == 1) return new int[]{0};
        int[] prefix = new int[n - 1];
        for (int i = 0; i < n - 1; i++) {
            prefix[i] = commonPrefix(words[i], words[i+1]);
        }
        int max = 0;
        for (int p : prefix) max = Math.max(max, p);
        int[] res = new int[n];
        for (int i = 0; i < n; i++) {
            int curMax;
            if (n == 2) {
                // Only one pair, removing either leaves no pairs
                res[i] = 0;
                continue;
            }
            if (i == 0) {
                // Remove first; pairs from 1 to n-2
                curMax = 0;
                for (int j = 1; j < n - 1; j++) curMax = Math.max(curMax, prefix[j]);
            } else if (i == n - 1) {
                // Remove last; pairs from 0 to n-3
                curMax = 0;
                for (int j = 0; j < n - 2; j++) curMax = Math.max(curMax, prefix[j]);
            } else {
                // Remove in the middle
                int combined = commonPrefix(words[i-1], words[i+1]);
                curMax = 0;
                for (int j = 0; j < n-1; j++) {
                    if (j == i-1) {
                        // This pair is replaced
                        curMax = Math.max(curMax, combined);
                    } else if (j == i) {
                        // This pair is removed
                        continue;
                    } else {
                        curMax = Math.max(curMax, prefix[j]);
                    }
                }
            }
            res[i] = curMax;
        }
        return res;
    }
}
# @lc code=end

File: 3598.longest-common-prefix-between-adjacent-strings-after-removals.go
#
# @lc app=leetcode id=3598 lang=golang
#
# [3598] Longest Common Prefix Between Adjacent Strings After Removals
#

# @lc code=start
func longestCommonPrefix(words []string) []int {
    n := len(words)
    lcp := func(a, b string) int {
        l := 0
        if len(a) > len(b) {
            a = a[:len(b)]
        } else if len(b) > len(a) {
            b = b[:len(a)]
        }
        for l < len(a) && a[l] == b[l] {
            l++
        }
        return l
    }
    if n == 1 {
        return []int{0}
    }
    prefixLens := make([]int, n-1)
    for i := 0; i < n-1; i++ {
        prefixLens[i] = lcp(words[i], words[i+1])
    }
    ans := make([]int, n)
    for i := 0; i < n; i++ {
        if n == 2 {
            ans[i] = 0
            continue
        }
        maxL := 0
        if i == 0 {
            // Remove first: pairs from (1,2), (2,3), ...
            for j := 1; j < n-1; j++ {
                if prefixLens[j] > maxL {
                    maxL = prefixLens[j]
                }
            }
        } else if i == n-1 {
            // Remove last: pairs from (0,1), ... (n-3,n-2)
            for j := 0; j < n-2; j++ {
                if prefixLens[j] > maxL {
                    maxL = prefixLens[j]
                }
            }
        } else {
            // Remove in the middle: (0,1), ..., (i-2,i-1), (i-1,i+1), (i+1,i+2), ...
            merged := lcp(words[i-1], words[i+1])
            for j := 0; j < n-2; j++ {
                if j == i-1 {
                    if merged > maxL {
                        maxL = merged
                    }
                } else {
                    idx := j
                    if j >= i {
                        idx = j+1 // Since one index is removed, shift
                    }
                    if prefixLens[idx] > maxL {
                        maxL = prefixLens[idx]
                    }
                }
            }
        }
        ans[i] = maxL
    }
    return ans
}
# @lc code=end